### PR TITLE
[ws-manager-api] Minor cluster selection refactoring

### PR DESCRIPTION
**Pull request by geropl**
## Description
After sleeping another night over the cluster selection implementation I had two gripes with it:
 1. cluster authorization is a [more](https://github.com/gitpod-io/gitpod/compare/gpl/cluster-selection-permissions?expand=1#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59L98-L101) or [less](https://github.com/gitpod-io/gitpod/compare/gpl/cluster-selection-permissions?expand=1#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59L117-L121) an "optimization" of the precedence implementation.
 This is now [implemented in a central constraint](https://github.com/gitpod-io/gitpod/pull/8034/files#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59R140-R160), and [applied to every ClusterSet][(https://github.com/gitpod-io/gitpod/compare/gpl/cluster-selection-permissions?expand=1#diff-69bcfb2991d810011d47c337182d6096cde4f6646da291505e7ba082b80fc0e0R49-R54](https://github.com/gitpod-io/gitpod/pull/8034/files#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59R67-R76)).
 This has the additional benefit that precedence and authorization are nicely decoupled and can be tested more easily.
 
 2. `constraintNewWorkspaceCluster` was somewhat of a special case because it was the only permission relevant for precedence so far ([while filtering _all_ permissions](https://github.com/gitpod-io/gitpod/compare/gpl/cluster-selection-permissions?expand=1#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59L104)). This is now generalized into [`constraintHasPermissions`](https://github.com/gitpod-io/gitpod/compare/gpl/cluster-selection-permissions?expand=1#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59R113-R117). With the change above this means all permissions may serve as qualifier for authorization (match cluster & user), while also, independently can serve as marker for precedence.

I wrote this PR to express my thoughts, so happy to discuss if others have other views on how this should evolve! :upside_down_face: :basketball: 

## Related Issue(s)

Fixes #

## How to test
 - `yarn test`

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** and additionally assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by geropl** [comment]
Talked to @csweichel and adjusted accordingly. The "authorization" is now implemented as [`Constraint`](https://github.com/gitpod-io/gitpod/pull/8034/files#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59R140-R160) as well and applied on all WorkspaceClusterSets in a [central place](https://github.com/gitpod-io/gitpod/pull/8034/files#diff-a6b16d5a1259e7e5284f0194a3e49ab4481fe395723eacc8d4570e7a6bfb9b59R70-R76).

**Review by csweichel — APPROVED** [review]
Awesome - thank you for the simplification

**Review by jankeromnes — APPROVED** [review]
Cool, many thanks for adding tests! 💯 🏅

Already approved by @csweichel, so briefly skimmed and rubber-stamping. ✅

**Pull request merged**
